ros_buildfarm/common.py

```diff
@@ -656,6 +668,46 @@ def get_implicitly_ignored_package_names(cached_pkgs, explicitly_ignored_pkg_nam
     return ignored_pkg_names.difference(explicitly_ignored_pkg_names)
 
 
+def filter_blocked_dependent_package_names(cached_pkgs, failed_pkg_names):
+    """Return the list of packages that are missing and not blocked.
+
+    Return the list of packages that are missing that are not depending
+    on other missing packages.
+    """
+    # get direct dependencies from distro cache for each package
+    direct_dependencies = {}
+    for pkg_name in cached_pkgs:
+        direct_dependencies[pkg_name] = get_direct_dependencies(
+            pkg_name, cached_pkgs, cached_pkgs) or set([])
+
+    # find recursive downstream deps for all explicitly ignored packages
+    all_deps = get_downstream_package_names(
+            failed_pkg_names, direct_dependencies)
+    while True:
+        blocked_pkgs = get_downstream_package_names(
+            all_deps, direct_dependencies)
+        if blocked_pkgs - all_deps:
+            all_deps |= blocked_pkgs
+            continue
+        break
```

Comment:
The logic within this loop should certainly ensure that it terminates. The implicitly assumption is that each iteration `all_deps` is increasing and therefore at some point reaching `blocked_pkgs` and `break`. That should be explicitly check with an `assert` to ensure this doesn't loop indefinitely.

Reply:
This logic is guaranteed to terminate. 

Without going into the details of `get_downstream_package_names`, it returns a subset of the packages in the dependency tree defined in `direct_dependencies` which does not change throughout the loop. Making no assumptions about a valid structure of the dependency tree. It will return zero or more packages from the dependency tree.

If it returns zero packages, the loop breaks. 

The results of each iteration is compared to a list of previously detected dependencies, and will only continue to loop if it's returning one or more new packages to add to the list. If it returns a set of packages which are fully contained in the known list it will break out. Thus the worst case is that it returns 1 new package every time, and in that case it will only run for N cycles where N is the number of packages in the dependency tree. And that worse case will only happen if there's a purely linear dependency with N layers.

The case of a malformed dependency tree, such as one with loops, it will still exit because the dependencies returned will be in the existing list and thus terminate the iteration as soon as the loop cycle starts to repeat because there are no new packages.

In reality this will iterate N times where N is the depth of the dependency tree from the initial packages.

Reply:
Thanks for describing the logic in more detail. I still think add an assertion to ensure this is the case is the defensive way to write this loop.

Reply:
What are you asking to be asserted?

You mention this above:

> The implicit assumption is that in each iteration all_deps is growing and therefore at some point reaching blocked_pkgs and break the loop. That should be explicitly checked with an assert to ensure this doesn't loop indefinitely.

But this is not an implicit assumption that should be asserted about. It is explicitly checked in the if statement. If there are no new dependencies then the loop breaks.

Reply:
> If it returns zero packages, the loop breaks.

I guess I didn't see that the condition `blocked_pkgs - all_deps` is doing both at the same time:
* it ensures that if `blocked_pkgs` is empty that it `breaks` as well as
* it ensures that `blocked_pkgs` later added to `all_deps` contains at least one element not already in `all_deps` which ensured progress.